### PR TITLE
AMRCS-293 map simultaneous saver

### PR DESCRIPTION
**Pull request by sdswoops824**
Relates to AMRCS-293

https://github.com/LexxPluss/LexxAuto/pull/3295

### CODE REVIEW

File: map_server/src/map_saver.cpp [hunk 1]
```diff
@@ -44,23 +44,37 @@ class MapGenerator
...
+    MapGenerator(const std::string& mapname1, const std::string& mapname2, int threshold_occupied, int threshold_free)
+      : mapname1_(mapname1), mapname2_(mapname2), saved_map1_(false), saved_map2_(false), threshold_occupied_(threshold_occupied), threshold_free_(threshold_free)
     {
       ros::NodeHandle n;
-      ROS_INFO("Waiting for the map");
-      map_sub_ = n.subscribe("map", 1, &MapGenerator::mapCallback, this);
+      ROS_INFO("Waiting for the maps");
+
+      map_sub1_ = n.subscribe("map", 1, &MapGenerator::mapCallback1, this);
+      map_sub2_ = n.subscribe("map_hi_intensity", 1, &MapGenerator::mapCallback2, this);
```
Comment (remmaTech12) [inline map_server/src/map_saver.cpp]: ? does this work? i thought we dont need to change a source code if the remapping works as expected in [this PR](https://github.com/LexxPluss/LexxAuto/pull/3295/files). am i wrong?
Screenshot from 2024-06-10 15-20-55

Comment (sdswoops824) [reply map_server/src/map_saver.cpp]: Yes, it works. I have tested in on 63002.
Please check the comments [here](https://lexxpluss.atlassian.net/browse/AMRCS-293?focusedCommentId=21663).

**Review by kokosabu — COMMENTED** [review]
I don't think the current implementation can create a scan-only map.
This is because we are trying to generate two maps with one MapGenerator instance.

I think it would be better to create two MapGenerator instances and create one map in each instance.

**Comment by sdswoops824** [comment]
> I don't think the current implementation can create a scan-only map. This is because we are trying to generate two maps with one MapGenerator instance. I think it would be better to create two MapGenerator instances and create one map in each instance.

@kokosabu 

Do you mean we should be able to perform both of the following:
`rosrun map_server map_saver -scan klab_scan -scan_hi klab_scan_hi`
`rosrun map_server map_saver -scan klab_scan`

OR 

we should run the server twice as in
`rosrun map_server map_saver -scan klab_scan`
`rosrun map_server map_saver -scan_hi klab_scan_hi`

**Comment by sdswoops824** [comment]
@kokosabu With [this](https://github.com/LexxPluss/navigation/pull/90/commits/49a5c0b3daecd5d17a8138d29639b7215af441c6) commit, it was enabled to save either one or both maps using CLI as follows:
```
rosrun map_server map_saver -scan test_scan_only
rosrun map_server map_saver -scan_hi test_scan_hi_only
rosrun map_server map_saver -scan_hi test_scan_hi -scan test_scan
rosrun map_server map_saver -scan test_scan -scan_hi test_scan_hi 
```

**Comment by kokosabu** [comment]
@sdswoops824 
My meaning is for the following

> `rosrun map_server map_saver -scan klab_scan -scan_hi klab_scan_hi`
> `rosrun map_server map_saver -scan klab_scan`

I thought it would be nice to be able to enable map creation with just -scan.

**Review by kokosabu — APPROVED** [review]

File: map_server/src/map_saver.cpp [hunk 2]
```diff
@@ -44,23 +44,39 @@ class MapGenerator
...
       ros::NodeHandle n;
-      ROS_INFO("Waiting for the map");
-      map_sub_ = n.subscribe("map", 1, &MapGenerator::mapCallback, this);
+      ROS_INFO("Waiting for the maps");
+
+      map_sub1_ = n.subscribe("map", 1, &MapGenerator::mapCallback1, this);
+      if (!single_map_) {
+        map_sub2_ = n.subscribe("map_hi_intensity", 1, &MapGenerator::mapCallback2, this);
+      }
+    }
```
Comment (yakato-jun) [inline map_server/src/map_saver.cpp:57]: subscribeの引数を省略形使わずに、bindで渡してあげれば、callbackを2つに分けなくても大丈夫だと思います。
https://wiki.ros.org/roscpp/Overview/Publishers%20and%20Subscribers#Functor_Objects
https://www.boost.org/doc/libs/1_85_0/libs/bind/doc/html/bind.html

```
  public:
    MapGenerator(const std::string& mapname1, const std::string& mapname2, int threshold_occupied, int threshold_free, bool single_map)
      : mapname1_(mapname1), mapname2_(mapname2), saved_map1_(false), saved_map2_(false), threshold_occupied_(threshold_occupied), threshold_free_(threshold_free), single_map_(single_map)
    {
      ros::NodeHandle n;
      ROS_INFO("Waiting for the maps");

      map_sub1_ = n.subscribe("map", 1, boost::bind(&MapGenerator::mapCallback, this, _1, mapname1_, boost::ref(saved_map1_)));
      if (!single_map_) {
        map_sub2_ = n.subscribe("map_hi_intensity", 1, boost::bind(&MapGenerator::mapCallback, this, _1, mapname2_, boost::ref(saved_map2_)));
      }
    }

    void mapCallback(const nav_msgs::OccupancyGridConstPtr& map, const std::string mapname, bool& saved_map)
    {
      ROS_INFO("Received a %d X %d map @ %.3f m/pix",
               map->info.width,
               map->info.height,
               map->info.resolution);


      std::string mapdatafile = mapname + ".pgm";
      ROS_INFO("Writing map occupancy data to %s", mapdatafile.c_str());
      FILE* out = fopen(mapdatafile.c_str(), "w");
      if (!out)
      {
        ROS_ERROR("Couldn't save map file to %s", mapdatafile.c_str());
        return;
      }

      fprintf(out, "P5\n# CREATOR: map_saver.cpp %.3f m/pix\n%d %d\n255\n",
              map->info.resolution, map->info.width, map->info.height);
      for(unsigned int y = 0; y < map->info.height; y++) {
        for(unsigned int x = 0; x < map->info.width; x++) {
          unsigned int i = x + (map->info.height - y - 1) * map->info.width;
          if (map->data[i] >= 0 && map->data[i] <= threshold_free_) { // [0,free)
            fputc(254, out);
          } else if (map->data[i] >= threshold_occupied_) { // (occ,255]
            fputc(000, out);
          } else { //occ [0.25,0.65]
            fputc(205, out);
          }
        }
      }

      fclose(out);


      std::string mapmetadatafile = mapname + ".yaml";
      ROS_INFO("Writing map occupancy data to %s", mapmetadatafile.c_str());
      FILE* yaml = fopen(mapmetadatafile.c_str(), "w");


      /*
resolution: 0.100000
origin: [0.000000, 0.000000, 0.000000]
#
negate: 0
occupied_thresh: 0.65
free_thresh: 0.196

       */

      geometry_msgs::Quaternion orientation = map->info.origin.orientation;
      tf2::Matrix3x3 mat(tf2::Quaternion(
        orientation.x,
        orientation.y,
        orientation.z,
        orientation.w
      ));
      double yaw, pitch, roll;
      mat.getEulerYPR(yaw, pitch, roll);

      fprintf(yaml, "image: %s\nresolution: %f\norigin: [%f, %f, %f]\nnegate: 0\noccupied_thresh: 0.65\nfree_thresh: 0.196\n\n",
              mapdatafile.c_str(), map->info.resolution, map->info.origin.position.x, map->info.origin.position.y, yaw);

      fclose(yaml);

      ROS_INFO("Done\n");
      saved_map = true;
    }
```

**Review by yakato-jun — APPROVED** [review]

File: map_server/src/map_saver.cpp (hunk 2 above)
Comment (sdswoops824) [reply map_server/src/map_saver.cpp:57]: ありがとうございました。明日試してみます。

**Review by remmaTech12 — APPROVED** [review]
The code should be alright. Thus I approve this PR. But please consider later about using only remap to achieve this functionality without any code change. If that is viable, good to revert these code change. Thank you.

**Comment by sdswoops824** [comment]
Decided to go with existing version of submodule and remapping technique.

**Pull request closed without merging**